### DISCUSSION TITLE
Confirm Message is Sent in Chat Before Clearing Message Bar

### DISCUSSION THREAD

**Discussion by andria-dev** [Chatterino/chatterino2#4539] [Ideas]
## The Problem
Disclaimer: All information about how chat works on Twitch's side is information I've learned from other people on here (see https://github.com/Chatterino/chatterino2/issues/1213#issuecomment-522368740). I have not verified any of that information.

On Twitch's side, chats with lots of incoming messages will pause and resume their acceptance of messages in an attempt to limit the number of messages. In order for a chatter's message to be received and added to the chat, it has to be received by Twitch's server while chat is still accepting messages. Apparently, Twitch's front-end implementation of chat displays your message as if it was sent even when Twitch's servers drop the message. Chatterino doesn't do this; instead, Chatterino only displays messages that are actually in the Twitch chat. So when it attempts to send your message and Twitch isn't accepting messages at that moment, **your message is cleared from the message bar as if it was sent, but nothing happens in the chat**.

With that out of the way, it would also appear that Twitch _silently_ drops the messages, so this doesn't appear to be a simple bug on the front-end but rather a difference of implementation. I personally think that there is a better way to handle this than the ways that Twitch and Chatterino handle it though.

## Different Ways of Handling This Issue
- Established Ways
  - Twitch: send the message to the back-end, display the user's message and clear the message bar regardless of whether or not the message was accepted, and continue retrieving messages from the back-end presumably while filtering out the user's messages from the back-end to prevent duplication.
    - Pros: It feels seamless.
    - Cons: There isn't a way to know if anyone ever received your message or if you need to send it again.
  - Chatterino: send the message to Twitch's back-end, clear the message bar regardless of whether or not the message was accepted, and continue retrieving messages from Twitch's back-end.
    - Pros: You can check if your message was received or not either by seeing it in chat or with searching for it with <kbd>Ctrl</kbd>+<kbd>F</kbd>.
    - Cons: Twitch only pauses acceptance of messages when there are lots of incoming messages, so it can be difficult to visually see if your message was received since there are so many messages. This often forces you to <kbd>Ctrl</kbd>+<kbd>F</kbd> after every message you send.
- **New way**: Send the message to Twitch's back-end, display some kind of indicator that let's the user know their message is being sent, **wait a second to receive new messages from Twitch**, and then **check if the message is in the chat**. If the message isn't in chat, display some kind of indicator that let's the user know that their message was not received so that they can resend their message if they want to. If the message is in chat, clear the message bar and temporarily display some kind of indicator that let's the user know the message was received. Alternatively, a setting could be introduced that automatically retries to send the message X amount of times when it isn't in chat.
  - Pros: No guessing if your message was sent or not, no need to <kbd>Ctrl</kbd>+<kbd>F</kbd> for each message, easy to resend your message.
  - Cons: Possible false negatives when experiencing very high latency, may add a small but noticeable delay when trying to quickly spam messages (although if that is an issue queueing up messages to be sent could be a way around that).

If anyone has any other ideas on how to handle this, please feel welcome to share them.

[Here is an interactive visual state machine of this new idea.](https://stately.ai/registry/editor/20a71670-2ca3-4f9f-adb0-087b776ed681?machineId=812032ca-8d87-4d21-b7b0-56fd0d2b0288&mode=Simulate)

**Comment by LosFarmosCTL** [comment]
> and continue retrieving messages from the back-end presumably while filtering out the user's messages from the back-end to prevent duplication.

Twitch servers don't actually send your own message back to you on the same connection, what chatterino does is open up separate read and write connections, send messages on one connection and join channels on the other one.

As for confirming that messages were sent, there would theoretically be a way easier way, you can append a `client-nonce` IRC tag to your outgoing message, that gets relayed onto the actual PRIVMSG Twitch sends to connected clients. In that nonce-tag you can include some UUID etc. and verify exactly which message was sent and which one wasn't. That tag is not documented though, so it would be relying on something Twitch could theoretically remove at any point.

Also, the main issue with implementing something like that is making it actually work well and not a pain to use, from a UI+UX standpoint. I can already imagine a lot of edge cases that would have to be handled just after thinking about it for a few minutes.

**Reply by andria-dev** [reply]
@LosFarmosCTL I mean, I feel like from the UI+UX standpoint, a simple state machine would handle it. Obviously, I don't have a very good understanding of Twitch API or the Chatterino codebase, but I feel like the idea I laid out above covers most cases. [Here is a visual state machine of my idea:](https://stately.ai/registry/editor/20a71670-2ca3-4f9f-adb0-087b776ed681?machineId=812032ca-8d87-4d21-b7b0-56fd0d2b0288&mode=Simulate)

![Chatterino Message Confirmation state machine](https://user-images.githubusercontent.com/19195374/232105913-d4975e20-5661-4f0c-9a2d-8ec54b33694f.png)

**Reply by pajlada** [reply]
The big challenges I see are:
1) How can we confidently confirm the message has come in?
  If we use the message body, what happens if Twitch transforms something in our message? This has happened before (removing double spaces, filtering out certain unicode characters)
  If we use the (seemingly undocumented) `client-nonce` tag, how will the UX suffer if we suddenly don't receive this message back?
2) How can we make sure the UX is an improvement in the usual case (message sent & received within 500ms) AND the various edge cases (message sent but not received, or message sent & received after 2s/4s/6s)?
3) Async code gets complex quickly, and having multiple checkers waiting with timers. How can we make sure an implementation that solves the above-mentioned issues doesn't become unmaintainable?

Thank you for putting effort into explaining your vision for it - I would be happy to support this functionality but for the effort of implementing it I can't put any sort of priority into implementing this. If we get a way to send messages through Helix with proper confirmation of it arriving, it would be a lot easier to support since we wouldn't need to do any guessing.

**Reply by Mm2PL** [reply]
~~`client-nonce` also has the problem that it just doesn't work for `/me` messages (or didn't use to).~~ This will be pretty complex, the only way to implement this is probably making a new singleton or new part of ...IrcServer especially for this. As you said it's hard to make this maintainable and keeping it in about one piece will be the hard part imo.

**Reply by treejadey** [reply]
UX/UI wise, it should probably work like this (please forgive my poor manual keyframe animation skills)
![chatterino_message](https://user-images.githubusercontent.com/18615981/232143135-acc507b4-8f70-479d-9b87-be94078b3c92.gif)

**Reply by andria-dev** [reply]
@pajlada 
> 1) How can we confidently confirm the message has come in?
>   If we use the message body, what happens if Twitch transforms something in our message? This has happened before (removing double spaces, filtering out certain unicode characters)

For a simpler approach, instead of using the message body, we could look at the sender of the message. If we just got a new message and it's our message, we confirm it. The only thing that this wouldn't support is 2 ppl chatting on the same account in the same channel at the same second on different clients, but that seems like a weird and unusual way to use Twitch.

**Reply by andria-dev** [reply]
@pajlada 
> 2) How can we make sure the UX is an improvement in the usual case (message sent & received within 500ms) AND the various edge cases (message sent but not received, or message sent & received after 2s/4s/6s)?

I'm not sure how someone is watching Twitch with 6s latency for chat alone, but I get your point. I think a potential solution to this could be to have an average latency timer that tracks how long it takes for each network request to be completed. With the average latency, we can then determine how much time to spend waiting for a response just after you send your message. And if the latency is very high (6s or more perhaps) we could disable this feature or at least offer to.

For example, if it took 2.5s on average for each of the last several network requests to complete. Maybe we wait 3s for confirmation. If the average is 0.5s, we might wait 1s for the confirmation. Over 6s, default to send and forget.

Also, with that, if this feature does ever pick up traction, making it optional (opt-out or opt-in idk) would be good and also fairly simple since disabling it would be just not waiting for a confirmation or anything. I'll update the state machine to try to reflect these changes when I get back to a computer

**Reply by leon-richardt** [reply]
> I'm not sure how someone is watching Twitch with 6s latency for chat alone, but I get your point. I think a potential solution to this could be to have an average latency timer that tracks how long it takes for each network request to be completed. With the average latency, we can then determine how much time to spend waiting for a response just after you send your message.

From anecdotal evidence, the latency that is experienced from TMI (the server-side implementation handling Twitch chat) is vastly different from what you would expect from a REST API. Instead, we could track the average time it takes for Twitch to bounce back our own messages to us and work with that value.

---

> How can we make sure the UX is an improvement in the usual case (message sent & received within 500ms) AND the various edge cases (message sent but not received, or message sent & received after 2s/4s/6s)?

I'm not quite convinced by this kind of argument. How often will these kind of edge cases realistically occur (they're called "edge cases" for a reason after all) and is it worth blocking an implementation over?

Also, as far as I am concerned, the UX in the "message sent but not received" case wouldn't be hurt. In the current state, the user would be under the impression that their message has not been sent successfully. In the proposed implementation, they would be told explicitly. From a "best-effort perspective", not much would change.

In the other mentioned case ("message sent & received after 2s/4s/6s"), we would warn a user prematurely. But again, the worst thing that could happen is an impatient user attempting to send the message again. However, an impatient user would likely already attempt to send their message again if it didn't show up after 2s/4s/6s in the current implementation.

**Comment by Felanbird** [comment]
I'd like to give my tin-foil hat opinion which is that chat messages will eventually be force-migrated to a sendChatMessage Helix API call, which _should_ at least offer us a 200 response. Though I have no proof to backup this statement, just assumptions based on the past migrations.

**Reply by Mm2PL** [reply]
> I'd like to give my tin-foil hat opinion which is that chat messages will eventually be force-migrated to a sendChatMessage Helix API call, which should at least offer us a 200 response.

I'd like to give my tin-foil hat opinion which is that this api if it ever exists, will be like the shitty whispers one which always returns 204.

**Comment by pajlada** [comment]
Would just showing an indicator for a little while when a message you sent is confirmed solve for the same issue? It would be much simpler implementation-wise


```
+-------------------------------------------+         +-----------------------------------------------------------+
| Message received where you are the sender | ------> | Pulse Message Confirmation Indicator for a short duration |
+-------------------------------------------+         +-----------------------------------------------------------+
```

**Reply by andria-dev** [reply]
@pajlada No I don't think that would be enough because the issue is really that there's no indication when your message **isn't** received. So there'd be no difference to what happens right now. And what happens right now is sending the message and searching chat for my message and re-sending multiple times in a row (for just one message)

**Reply by pajlada** [reply]
The reason I'm trying to avoid the `message isn't received` logic is because it's going to be impractical to implement since the API we use does not provide us with the necessary data, and doing guesswork **will** make for bad UX for some users.

If the user story is: "As a Twitch chatter, I want to make sure the important message I am sending is being sent", then you currently have the following options in Chatterino:
1. Enabling the "Your messages (automatic)" highlight under Highlights -> Users feature to color your message differently, so you have an easier time seeing your own messages in chat.
2. Adding a new split with a filter that only shows messages that you sent

My proposed additional solution is to highlight the input bar when a message of yours is received to give an additional highlight message that "a message of yours has been confirmed to be received by Twitch", which should solve for that user story.

**Reply by andria-dev** [reply]
@pajlada The visual example from @treuks seems to show one option for a decent UX implementation of this. https://github.com/Chatterino/chatterino2/discussions/4539#discussioncomment-5619843

Specifically, their UX implementation grays out the message you're sending while we're checking for confirmation. If it gets confirmed, we remove the message from the message bar. If it doesn't get confirmed in time, it simply doesn't get removed from the message bar. All of that done in a very short span of time. That seems like a pretty positive base UX experience for sending a chat.

Also, it seems like the current options that Chatterino users have (the ones you presented above) are very similar to the idea I've presented, the major difference being that the user does the manual work every time to look for the message instead of the application automating that for you. That being said, if the underlying processes for the above listed current options for Chatterino users are more desirable than the idea I presented, we could tailor the logic towards one of those.

For example, instead of only searching for the users message for x amount of time after sending it, we could continuously attempt to confirm the user's last message was received until we find it or they send another message. While continuously checking, we could present some kind of indicator that let's the user know we haven't confirmed the last message (perhaps keeping the last message grayed out just above the message bar with a re-send button).

**Reply by pajlada** [reply]
> Specifically, their UX implementation grays out the message you're sending while we're checking for confirmation. If it gets confirmed, we remove the message from the message bar. If it doesn't get confirmed in time, it simply doesn't get removed from the message bar. All of that done in a very short span of time. That seems like a pretty positive base UX experience for sending a chat.

This will lock up the message bar until the message has been confirmed sent - I think this will lead to a bad user experience.

> Also, it seems like the current options that Chatterino users have (the ones you presented above) are very similar to the idea I've presented, the major difference being that the user does the manual work every time to look for the message instead of the application automating that for you. 

Could you clarify which manual processes in the alternatives I mentioned above could be automated? The only automation I see is not having to press "arrow up" to re-send your last message if your message wasn't sent.

**Reply by andria-dev** [reply]
@pajlada 

> This will lock up the message bar until the message has been confirmed sent

The UX shown by @treuks uses a timeout to not lock up the message bar for longer than a second. It just doesn't clear the message bar if the message wasn't actually sent. From there you could type a new message or resend the last one. However, the idea I presented after that does give an entirely non-blocking option


> Could you clarify which manual processes

To clarify, the manual processes are 1. changing settings to make your own messages visually stand out more and looking through a fast moving chat to find your message you just sent or 2. creating a split for each channel, renaming the tab so it isn't doubled, creating a filter for your own messages, adding that filter to each of those splits, and then monitoring the filtered split to see if your message was sent. That second one is also very much not a simple process for a inexperienced Chatterino user (I didn't even think of that until you explained it).

At a top-level view, what I mean by "automating manual processes" is instead of creating ways to make your own messages stand out in the settings or with splits+filters that aren't on by default, we can create a similar non-blocking process where Chatterino *by default* distinguishes which messages are yours and then confirms if you received your message back all while presenting this information to you in a way that ensures you won't miss what happened to your message or miss anything else happening in chat. The more detailed explanation of how this could work being in my previous message of course.

**Reply by pajlada** [reply]
The new state machine you linked in the top post (https://stately.ai/registry/editor/20a71670-2ca3-4f9f-adb0-087b776ed681?mode=Simulate&machineId=812032ca-8d87-4d21-b7b0-56fd0d2b0288) looks good to me :+1:
I would caution against adding the more advanced functionality in an initial PR (like the 'send last message' helper) to keep things simple.

Timing will require actual implementation to experiment what feels right, and we'd need to figure out what place makes most sense to put this (circle indicator in the message box, glow on the message box or something) but I would accept this functionality as a PR.

Thank you for being patient and for showing us that state machine website!
